### PR TITLE
Add Provider and QUERY tracepoints

### PR DESCRIPTION
**Pull request by nhorman**
A take away from the webinar as well as a request as part of our PQ implementation was to support provider load and query tracing as part of our trace api.  This PR provides that, allowing users to specify OPENSSL_TRACE=[PROVIDER|QUERY] as an environment variable to log trace messages during those processes

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/property/property.c [hunk 1]
```diff
@@ -324,7 +358,21 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
         OPENSSL_free(impl);
         return 0;
     }
+
+    /*
+     * flush the method store of any implementation that already exists
```
Comment (paulidale) [inline crypto/property/property.c:363]: Capital _F_ definitely.

File: crypto/provider_core.c [hunk 2]
```diff
@@ -1612,59 +1617,266 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
+ */
+static void trace_print_param_values(const OSSL_PARAM p[], BIO *b)
+{
+    int64_t i;
+    uint64_t u;
+
+    for (; p->key != NULL; p++) {
+        BIO_printf(b, "%s: ", p->key);
+        switch (p->data_type) {
+        case OSSL_PARAM_UNSIGNED_INTEGER:
```
Comment (paulidale) [inline crypto/provider_core.c:1645]: `OSSL_PARAM_REAL` is not handled here.  It needs to be conditioned on `OPENSSL_SYS_UEFI` which doesn't have real support.

File: crypto/property/property.c [hunk 3]
```diff
@@ -355,9 +407,28 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
             && tmpimpl->properties == impl->properties)
             break;
     }
+
+    /*
+     * Check the number of implementation on the impl stack.
+     * If it matches the number we counted above, and the push
+     * operation is successful, then we're done here, and can return
+     * success.
+     * NOTE: This seems ripe for improvement, in that if we find a duplicate
```
Comment (paulidale) [inline crypto/property/property.c:416]: Do we really care enough to ever introduce this optimisation?
I don't think this will be common enough to justify complicating the code.

Give the opening phrase, it's likely somebody will make such a change based on good intentions alone.

IMO, this part of the comment should be removed.  Optionally, make the suggested change but I'm not going to push for that.

File: crypto/property/property.c [hunk 4]
```diff
@@ -298,11 +319,24 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
     int ret = 0;
     int i;
 
+    /*
+     * fail if someone is trying:
```
Comment (paulidale) [inline crypto/property/property.c:323]: Capital _F_ perhaps?

**Review by paulidale — CHANGES_REQUESTED** [review]

File: crypto/property/property.c (hunk 3 above)
Comment (nhorman) [reply crypto/property/property.c:416]: I don't think it would be particularly complicated.  Its the difference between looping over all the providers once, and jumping to err if a match was found, then doing a push, versus breaking the search loop early, and doing a comparison to an extra call to IMPLEMENTATION_num.  Its probably not significant in terms of performance, but it would make the code more understandable and compact.  The only reason I didn't just do it here was the desire to avoid feature creep in the PR

File: crypto/property/property.c [hunk 5]
```diff
@@ -549,6 +685,13 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store,
         }
     }
 
+    /*
+     * Search for a provider that provides this implementaiton.
```
Comment (slontis) [inline crypto/property/property.c:689]:
```suggestion
     * Search for a provider that provides this implementation.
```

File: crypto/provider_core.c [hunk 6]
```diff
@@ -1612,50 +1618,255 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
+ * - OSSL_PARAM_UTF8_PTR: Pointer to a UTF-8 string.
+ * - OSSL_PARAM_UTF8_STRING: UTF-8 string.
+ * - OSSL_PARAM_OCTET_PTR/OCTET_STRING: Binary data (size in bytes).
+ */
+#ifndef FIPS_MODULE
+static void trace_print_param_values(const OSSL_PARAM p[], BIO *b)
+{
+    int64_t i;
+    uint64_t u;
+    double d;
```
Comment (paulidale) [inline crypto/provider_core.c:1643]: This will need the #ifdef guard too or compilers will generate unused variable warnings.
Alternatively, declare it in the case block.

**Review by paulidale — APPROVED** [review]

File: crypto/property/property.c [hunk 7]
```diff
@@ -288,6 +289,26 @@ static int ossl_method_store_insert(OSSL_METHOD_STORE *store, ALGORITHM *alg)
...
+/**
+ * @brief Adds a method to the specified method store.
+ *
+ * This function adds a new method to the provided method store, associating it
+ * with a specified NID, properties, and provider. The method is stored with
+ * reference count and destruction callbacks.
+ *
+ * @param store Pointer to the OSSL_METHOD_STORE where the method will be added.
+ * @param prov Pointer to the OSSL_PROVIDER representing the provider of the method.
+ * @param nid NID (numerical identifier) associated with the method.
```
Comment (levitte) [inline crypto/property/property.c:301]: It's not a NID in the sense of the `NID_*` macros.  The parameter is not well named, I would just have named it `id` or `internal_id`.

File: crypto/property/property.c [hunk 8]
```diff
@@ -298,11 +319,24 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
     int ret = 0;
     int i;
 
+    /*
+     * Fail if someone is trying:
+     * 1) To add the NID_undef nid (0)
```
Comment (levitte) [inline crypto/property/property.c:324]: No, please don't conflate this with `NID_*` macros

File: crypto/property/property.c [hunk 9]
```diff
@@ -504,6 +592,25 @@ void ossl_method_store_do_all(OSSL_METHOD_STORE *store,
...
 
+/**
+ * @brief Fetches a method from the method store matching the given properties.
+ *
+ * This function searches the method store for an implementation of a specified
+ * method, identified by its NID, and matching the given property query. If
+ * successful, it returns the method and its associated provider.
+ *
+ * @param store Pointer to the OSSL_METHOD_STORE from which to fetch the method.
+ * @param nid NID (numerical identifier) of the method to be fetched.
```
Comment (levitte) [inline crypto/property/property.c:603]: Same here

File: crypto/property/property.c [hunk 10]
```diff
@@ -516,6 +623,10 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store,
     int ret = 0;
     int j, best = -1, score, optional;
 
+    /*
+     * Typical check, fail if we try to lookup NID_undef, a null method
```
Comment (levitte) [inline crypto/property/property.c:627]: Nope, not `NID_undef`, just zero or less

**Review by levitte — CHANGES_REQUESTED** [review]
Please don't conflate the identities used here with NIDs

File: test/trace_api_test.c [hunk 11]
```diff
@@ -27,25 +27,27 @@ static int test_trace_categories(void)
...
+            CASE(X509V3_POLICY);
+            CASE(BN_CTX);
+            CASE(CMP);
+            CASE(STORE);
+            CASE(DECODER);
+            CASE(ENCODER);
+            CASE(REF_COUNT);
+            CASE(HTTP);
+            CASE(PROVIDER);
+            CASE(QUERY);
```
Comment (levitte) [inline test/trace_api_test.c:50]: This indentation makes no sense when you look at the macro.  That, or all `case ...:` should should be indented.  I understand that the check format script tells you something else, but...

File: crypto/property/property.c [hunk 12]
```diff
@@ -288,6 +289,29 @@ static int ossl_method_store_insert(OSSL_METHOD_STORE *store, ALGORITHM *alg)
...
+ * @param method Pointer to the method to be added.
+ * @param method_up_ref Function pointer for incrementing the method ref count.
+ * @param method_destruct Function pointer for destroying the method.
+ *
+ * @return 1 if the method is successfully added, 0 on failure.
+ *
+ * If tracing is enabled, a message is printed indicating that the method is
+ * being added to the method store.
+ *
+ * NOTE: The nid parameter here is _not_ a nid in the sense of the NID_* macros.
```
Comment (baentsch) [inline crypto/property/property.c:312]: Then wouldn't it be much better to call this "mid" or "meth_id"?

File: crypto/property/property.c [hunk 13]
```diff
@@ -298,11 +322,24 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
     int ret = 0;
     int i;
 
+    /*
+     * Fail if someone is trying:
+     * 1) To add an undefined identifier nid (0)
+     * 2) To add an empty method
+     * 3) To add to a non-existant store
```
Comment (baentsch) [inline crypto/property/property.c:329]: nit: ...non-existent...

File: crypto/property/property.c [hunk 14]
```diff
@@ -560,6 +698,12 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store,
         }
         goto fin;
     }
+
+    /*
+     * If there are optional properties specified
+     * the search again, and select the provider that matches the
```
Comment (baentsch) [inline crypto/property/property.c:704]: nit: "the" -> "then"

File: crypto/provider_core.c [hunk 15]
```diff
@@ -964,6 +964,12 @@ static int provider_init(OSSL_PROVIDER *prov)
                        prov->name);
         goto end;
     }
+#ifndef FIPS_MODULE
```
Comment (baentsch) [inline crypto/provider_core.c:967]: Why are trace statements above/before not "FIPS-#ifdef guarded"?

File: crypto/provider_core.c [hunk 16]
```diff
@@ -1612,50 +1618,257 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
 const OSSL_PARAM *ossl_provider_gettable_params(const OSSL_PROVIDER *prov)
 {
-    return prov->gettable_params == NULL
-        ? NULL : prov->gettable_params(prov->provctx);
+    const OSSL_PARAM *ret = NULL;
+
+    if (prov->gettable_params != NULL) {
+        ret = prov->gettable_params(prov->provctx);
+#ifndef FIPS_MODULE
+        OSSL_TRACE_BEGIN(PROVIDER) {
```
Comment (baentsch) [inline crypto/provider_core.c:1738]: No trace is generated if provider doesn't implement "gettable_params". Intentional?

File: crypto/provider_core.c [hunk 17]
```diff
@@ -1612,50 +1618,257 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
 {
     int ret;
 
     if (prov->self_test == NULL)
         return 1;
+
     ret = prov->self_test(prov->provctx);
+
+#ifndef FIPS_MODULE
+    OSSL_TRACE_BEGIN(PROVIDER) {
```
Comment (baentsch) [inline crypto/provider_core.c:1806]: Again, no output if no self-test present...

**Review by baentsch — COMMENTED** [review]
@nhorman Looks mostly logical to me but I'm hardly in the position to approve or not. So, see some minor nits, some questions mostly for my education (feel free to resolve these) and one more serious question: Why is the term nid used here? It's pretty overloaded for users of the OpenSSL API... Could this be changed?

File: crypto/provider_core.c [hunk 18]
```diff
@@ -1612,50 +1618,257 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
+ *
+ * The supported parameter types are:
+ * - OSSL_PARAM_UNSIGNED_INTEGER: Unsigned integer values.
+ * - OSSL_PARAM_INTEGER: Signed integer values.
+ * - OSSL_PARAM_UTF8_PTR: Pointer to a UTF-8 string.
+ * - OSSL_PARAM_UTF8_STRING: UTF-8 string.
+ * - OSSL_PARAM_OCTET_PTR/OCTET_STRING: Binary data (size in bytes).
+ */
+#ifndef FIPS_MODULE
+static void trace_print_param_values(const OSSL_PARAM p[], BIO *b)
```
Comment (t-j-h) [inline crypto/provider_core.c:1639]: This is a missing function from the library itself - we have a couple of places where we want to show a user the details of an OSSL_PARAM array and that should be in the base API - i.e. this function should be public - and should take a flag that allows for different formats to be specified in future with this being the "default" and the other usages of this sort of thing. We should have a version which reports the type for each.

If we are going to add this in here without going that I suggest an issue to sort that out be added.
Basically anywhere where you are querying what is supported and want to tell the user (i,.e. EVP_PKEY_gettable_params output) you shouldn't have to repeat this sort of code.

File: crypto/provider_core.c [hunk 19]
```diff
@@ -1612,50 +1618,257 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
+            BIO_printf(b, "'%s'\n", *(char **)(p->data));
+            break;
+        case OSSL_PARAM_UTF8_STRING:
+            BIO_printf(b, "'%s'\n", (char *)p->data);
+            break;
+        case OSSL_PARAM_OCTET_PTR:
+        case OSSL_PARAM_OCTET_STRING:
+            BIO_printf(b, "<%zu bytes>\n", p->data_size);
+            break;
+# ifndef OPENSSL_SYS_UEFI
```
Comment (t-j-h) [inline crypto/provider_core.c:1672]: I wouldn't handle this around this way - the type is known - someone can elect to pass it in - the platform doesn't have the conversion support routines - but it is a known type - handled this way it is seen as unknown - and it isn't - it is unsupported on this platform. Look at the other usage where OSSL_PARAM_REAL is on the "outside" and the ifdef check is on the "inside".

File: crypto/provider_core.c (hunk 16 above)
Comment (t-j-h) [reply crypto/provider_core.c:1738]: It should trace that it does not exist IMHO ...

File: crypto/provider_core.c (hunk 17 above)
Comment (t-j-h) [reply crypto/provider_core.c:1806]: It should trace if not present IMHO

File: test/trace_api_test.c [hunk 20]
```diff
@@ -17,48 +17,67 @@ static int test_trace_categories(void)
...
     for (cat_num = -1; cat_num <= OSSL_TRACE_CATEGORY_NUM + 1; ++cat_num) {
         const char *cat_name = OSSL_trace_get_category_name(cat_num);
-        int is_cat_name_eq = 0;
+        const char *expected_cat_name = NULL;
         int ret_cat_num;
-        int expected_ret;
 
+#define SET_EXPECTED_CAT_NAME(name) expected_cat_name = #name; break
         switch (cat_num) {
-#define CASE(name) \
```
Comment (t-j-h) [inline test/trace_api_test.c:25]: I really don't get why there is a refactoring in here ... it seems totally unrelated to the rest of the PR

File: crypto/property/property.c [hunk 21]
```diff
@@ -582,6 +726,18 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store,
...
+
+    OSSL_TRACE_BEGIN(QUERY) {
+        char buf[512];
+        int size;
+
+        size = ossl_property_list_to_string(NULL, pq, buf, 512);
+        BIO_printf(trc_out, "method store query with properties %s "
+                   "resolves to provider %s\n",
+                   size == 0 ? "none" : buf,
+                   ossl_provider_name(best_impl->provider));
```
Comment (Sashan) [inline crypto/property/property.c:738]: How do we know `best_impl` is not NULL here? In other words: my concern is we may be running here even if we found nothing.

I think code here should be moved to else branch at line 723 or we should re-check local variable `ret`. If ret is zero then trace should say no provider matching query could be found.

File: crypto/provider_core.c (hunk 18 above)
Comment (Sashan) [reply crypto/provider_core.c:1639]: > This is a missing function from the library itself - we have a couple of places where we want to show a user the details of an OSSL_PARAM array and that should be in the base API - i.e. this function should be public - and should take a flag that allows for different formats to be specified in future with this being the "default" and the other usages of this sort of thing. We should have a version which reports the type for each.
</snip>

I prefer to get this done this PR and then address other places separately. this will be the right time to promote this function to `ossl_trace_param_values()`

**Review by Sashan — CHANGES_REQUESTED** [review]
Looks reasonable, but some finishing touches are required. thanks.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Comment by nhorman** [comment]
@baentsch I think just changing a parameter name might not be considered abi breaking, but I'm not sure what our tests would say about it.  Probably best to do in a separate pr

File: test/trace_api_test.c (hunk 20 above)
Comment (nhorman) [reply test/trace_api_test.c:25]: It was the result of coding style errors, which I just happened to fix up here

**Review by Sashan — APPROVED** [review]
changes look good to me, thanks.

File: crypto/params_from_text.c [hunk 22]
```diff
@@ -197,6 +199,139 @@ static int construct_from_text(OSSL_PARAM *to, const OSSL_PARAM *paramdef,
...
+ *                are printed.
+ *
+ * This function iterates through the given array of OSSL_PARAM structures,
+ * printing each key to an in-memory buffer, and optionally printing its
+ * value based on the provided data type. Supported types include integers,
+ * strings, octet strings, and real numbers.
+ *
+ * Return:    The size of the buffer on success, or -1 on failure.
+ */
+int OSSL_PARAM_print_to_buf(const OSSL_PARAM p[], char **buf, int print_values)
```
Comment (mattcaswell) [inline crypto/params_from_text.c:219]: Would `ossl_ssize_t` be a better return value since this represents a size?

File: crypto/params_from_text.c [hunk 23]
```diff
@@ -197,6 +199,139 @@ static int construct_from_text(OSSL_PARAM *to, const OSSL_PARAM *paramdef,
...
+     */
+    for (; p->key != NULL; p++) {
+        ok = -1;
+        ok = BIO_printf(b, "%s: ", p->key);
+
+        /*
+         * if printing of values was not requested, just move on
+         * to the next param, after adding a newline to the buffer
+         */
+        if (print_values == 0) {
```
Comment (mattcaswell) [inline crypto/params_from_text.c:247]: Shouldn't we check the value of ok here?

File: crypto/params_from_text.c [hunk 24]
```diff
@@ -197,6 +199,139 @@ static int construct_from_text(OSSL_PARAM *to, const OSSL_PARAM *paramdef,
...
+                ok = BIO_printf(b, "%llu\n", (unsigned long long int)u);
+            else
+                ok = BIO_printf(b, "error getting value\n");
+            break;
+        case OSSL_PARAM_INTEGER:
+            if (OSSL_PARAM_get_int64(p, &i))
+                ok = BIO_printf(b, "%lld\n", (long long int)i);
+            else
+                ok = BIO_printf(b, "error getting value\n");
+            break;
```
Comment (mattcaswell) [inline crypto/params_from_text.c:264]: This doesn't seem like it will handle BIGNUMs very well.

File: crypto/params_from_text.c [hunk 25]
```diff
@@ -197,6 +199,139 @@ static int construct_from_text(OSSL_PARAM *to, const OSSL_PARAM *paramdef,
...
+        case OSSL_PARAM_UTF8_PTR:
+            ok = BIO_printf(b, "'%s'\n", *(char **)(p->data));
+            break;
+        case OSSL_PARAM_UTF8_STRING:
+            ok = BIO_printf(b, "'%s'\n", (char *)p->data);
+            break;
+        case OSSL_PARAM_OCTET_PTR:
+        case OSSL_PARAM_OCTET_STRING:
+            ok = BIO_printf(b, "<%zu bytes>\n", p->data_size);
```
Comment (mattcaswell) [inline crypto/params_from_text.c:273]: Consider using `BIO_dump` to print the octet string contents

File: crypto/params_from_text.c (hunk 22 above)
Comment (mattcaswell) [inline crypto/params_from_text.c:219]: I'm wondering why we have chosen to print to a buf, rather than print to a BIO which seems more natural and consistent with other OpenSSL APIs.

File: doc/man3/OSSL_PARAM_print_to_buf.pod [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+and on succesful return, must be freed by the caller. B<print_values> is a
+control parameter, indicating that key values should be printed, in addition to
+key names.
+
+=head1 RETURN VALUES
+
+OSSL_PARAM_print_to_buf() returns the length of B<*buf> on success, and -1
+if any error has occured.
+
+=head1 COPYRIGHT
```
Comment (mattcaswell) [inline doc/man3/OSSL_PARAM_print_to_buf.pod:32]: We should have a HISTORY section saying that this was added in 3.5

File: doc/man3/OSSL_PARAM_print_to_buf.pod [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+                             int print_values);
+
+=head1 DESCRIPTION
+
+OSSL_PARAM_print_to_buf() formats each parameter contained in the
+passed in array of OSSL_PARAM values B<p>, and prints both the key,
+and optionally its value, to an in memory buffer, returning that buffer
+to the caller.  B<p> must be a non-null array of OSSL_PARAM values, terminated
+with a value containing a null B<key> member.  B<buf> is a pointer to a
+character array.  B<*buf> is automatically sized an allocated by this routine,
```
Comment (mattcaswell) [inline doc/man3/OSSL_PARAM_print_to_buf.pod:22]: typo: *and* allocated

File: crypto/property/property.c [hunk 28]
```diff
@@ -288,6 +289,31 @@ static int ossl_method_store_insert(OSSL_METHOD_STORE *store, ALGORITHM *alg)
...
 
+/**
+ * @brief Adds a method to the specified method store.
+ *
+ * This function adds a new method to the provided method store, associating it
+ * with a specified id, properties, and provider. The method is stored with
+ * reference count and destruction callbacks.
+ *
+ * @param store Pointer to the OSSL_METHOD_STORE where the method will be added.
+ *        must be non-null.
```
Comment (mattcaswell) [inline crypto/property/property.c:300]: Line this up with "Pointer" on the previous line? Also, make the sentence start with a capital.

File: crypto/property/property.c [hunk 29]
```diff
@@ -353,11 +402,19 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
 
         if (tmpimpl->provider == impl->provider
             && tmpimpl->properties == impl->properties)
-            break;
+            goto err;
```
Comment (mattcaswell) [inline crypto/property/property.c:405]: This is not correct. It will cause `alg_cleanup` to be called that will free the alg. But it is still in use.

File: crypto/property/property.c [hunk 30]
```diff
@@ -504,6 +589,29 @@ void ossl_method_store_do_all(OSSL_METHOD_STORE *store,
...
 
+/**
+ * @brief Fetches a method from the method store matching the given properties.
+ *
+ * This function searches the method store for an implementation of a specified
+ * method, identified by its id (nid), and matching the given property query. If
+ * successful, it returns the method and its associated provider.
+ *
+ * @param store Pointer to the OSSL_METHOD_STORE from which to fetch the method.
+ *        Must be non-null
```
Comment (mattcaswell) [inline crypto/property/property.c:600]: Line up with "Pointer"?

File: crypto/property/property.c [hunk 31]
```diff
@@ -549,6 +667,13 @@ int ossl_method_store_fetch(OSSL_METHOD_STORE *store,
         }
     }
 
+    /*
+     * Search for a provider that provides this implementation.
+     * if the requested provider is NULL, then any provider will do,
```
Comment (mattcaswell) [inline crypto/property/property.c:672]: Nit: s/if/If/

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_core.c (hunk 15 above)
Comment (nhorman) [reply crypto/provider_core.c:967]: There was one that was missed, I'm fixing it

File: crypto/property/property.c (hunk 12 above)
Comment (nhorman) [reply crypto/property/property.c:314]: Perhaps, but that seems outside the scope of this PR to me

File: crypto/property/property.c (hunk 28 above)
Comment (nhorman) [reply crypto/property/property.c:300]: I'll do the alignment, but the sentence starts with the word Pointer, which is capitalized.  store is the name of the parameter which is specified in lowercase.

File: crypto/params_from_text.c (hunk 22 above)
Comment (nhorman) [reply crypto/params_from_text.c]: no longer needed with conversion to print_to_bio

**Review by Sashan — CHANGES_REQUESTED** [review]
looks good, there is a conflict to be resolved in `util/libcrypto.num`.

File: util/libcrypto.num [hunk 32]
```diff
@@ -5750,3 +5750,4 @@ i2d_OSSL_ROLE_SPEC_CERT_ID_SYNTAX       ?	3_5_0	EXIST::FUNCTION:
 OSSL_ROLE_SPEC_CERT_ID_SYNTAX_free      ?	3_5_0	EXIST::FUNCTION:
 OSSL_ROLE_SPEC_CERT_ID_SYNTAX_new       ?	3_5_0	EXIST::FUNCTION:
 OSSL_ROLE_SPEC_CERT_ID_SYNTAX_it        ?	3_5_0	EXIST::FUNCTION:
```
[the remark this answers is not in the document]
Comment (nhorman) [reply util/libcrypto.num]: rebased to resolve conflicts

**Review by Sashan — APPROVED** [review]
looks good to me now. thanks.

**Review by Sashan — APPROVED** [review]
my approval holds. thanks.

File: crypto/property/property.c [hunk 33]
```diff
@@ -355,9 +404,17 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
             && tmpimpl->properties == impl->properties)
             break;
     }
-    if (i == sk_IMPLEMENTATION_num(alg->impls)
-        && sk_IMPLEMENTATION_push(alg->impls, impl))
```
Comment (mattcaswell) [inline crypto/property/property.c:359]: Why this change? It looks like we are now unconditionally pushing to the stack regardless of whether there is one there or not - which was the whole point of the previous loop?

File: crypto/property/property.c [hunk 34]
```diff
@@ -426,9 +498,22 @@ alg_cleanup_by_provider(ossl_uintmax_t idx, ALGORITHM *alg, void *arg)
         IMPLEMENTATION *impl = sk_IMPLEMENTATION_value(alg->impls, i);
 
         if (impl->provider == data->prov) {
-            impl_free(impl);
+
+            OSSL_TRACE_BEGIN(QUERY) {
+                char buf[512];
+                size_t size;
+
+                size = ossl_property_list_to_string(NULL, impl->properties, buf, 512);
```
Comment (mattcaswell) [inline crypto/property/property.c:506]:
```suggestion
                size = ossl_property_list_to_string(NULL, impl->properties, buf, sizeof(buf));
```

File: crypto/provider_core.c [hunk 35]
```diff
@@ -1612,59 +1618,247 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
-    return prov->get_params == NULL
-        ? 0 : prov->get_params(prov->provctx, params);
+    int ret;
+
+    if (prov->get_params == NULL)
+        return 0;
+
+    ret = prov->get_params(prov->provctx, params);
+#ifndef FIPS_MODULE
+    if (ret == 1) {
```
Comment (mattcaswell) [inline crypto/provider_core.c:1713]: Shouldn't we provide some output even if `get_params` fails?

File: crypto/provider_core.c [hunk 36]
```diff
@@ -1612,59 +1618,247 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
+            BIO_printf(trc_out, "provider %s lacks query operation!\n",
+                       prov->name);
+        } OSSL_TRACE_END(PROVIDER);
+#endif
         return NULL;
+    }
+
     res = prov->query_operation(prov->provctx, operation_id, no_cache);
+#ifndef FIPS_MODULE
+    if (res != NULL) {
```
Comment (mattcaswell) [inline crypto/provider_core.c:1840]: Shouldn't we provide some output even if query fails?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_core.c (hunk 36 above)
Comment (t8m) [inline crypto/provider_core.c:1840]: Should the failed query be traced?

File: crypto/provider_core.c (hunk 35 above)
Comment (t8m) [inline crypto/provider_core.c:1713]: Should the failure be traced?

File: crypto/provider_core.c [hunk 37]
```diff
@@ -1612,59 +1618,247 @@ OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
...
+                       ossl_provider_name(prov));
+            BIO_printf(trc_out, "Parameters:\n");
+            if (prov->gettable_params != NULL) {
+                if (!OSSL_PARAM_print_to_bio(ret, trc_out, 0))
+                    BIO_printf(trc_out, "Failed to parse param values\n");
+                OPENSSL_free(buf);
+            } else {
+                BIO_printf(trc_out, "Provider doesn't implement gettable_params\n");
+            }
+        } OSSL_TRACE_END(PROVIDER);
```
Comment (t8m) [inline crypto/provider_core.c:1684]: There is too much indentation here.

File: crypto/property/property.c [hunk 38]
```diff
@@ -355,9 +404,17 @@ int ossl_method_store_add(OSSL_METHOD_STORE *store, const OSSL_PROVIDER *prov,
             && tmpimpl->properties == impl->properties)
             break;
     }
-    if (i == sk_IMPLEMENTATION_num(alg->impls)
-        && sk_IMPLEMENTATION_push(alg->impls, impl))
+
+    if (sk_IMPLEMENTATION_push(alg->impls, impl)) {
```
Comment (t8m) [inline crypto/property/property.c:408]: You cannot remove the first condition - this way you'll insert duplicates.

File: crypto/params_from_text.c [hunk 39]
```diff
@@ -197,6 +199,111 @@ static int construct_from_text(OSSL_PARAM *to, const OSSL_PARAM *paramdef,
...
+ *                are printed.
+ *
+ * This function iterates through the given array of OSSL_PARAM structures,
+ * printing each key to an in-memory buffer, and optionally printing its
+ * value based on the provided data type. Supported types include integers,
+ * strings, octet strings, and real numbers.
+ *
+ * Return:    1 on success, 0 on failure.
+ */
+int OSSL_PARAM_print_to_bio(const OSSL_PARAM p[], BIO *bio, int print_values)
```
Comment (t8m) [inline crypto/params_from_text.c:217]: For some reason I prefer `*p` to `p[]` - it is also more consistent with other functions working with OSSL_PARAM arrays. Also in the man page you have `*p`.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by mattcaswell — DISMISSED** [review]

File: doc/man3/OSSL_PARAM_print_to_bio.pod [hunk 40]
```diff
@@ -15,11 +15,11 @@ OSSL_PARAM_print_to_bio
 =head1 DESCRIPTION
 
 OSSL_PARAM_print_to_bio() formats each parameter contained in the
-passed in array of OSSL_PARAM values B<p>, and prints both the key,
+passed in array of OSSL_PARAM values I<p>, and prints both the key,
 and optionally its value, to a provided BIO.
-B<p> must be a non-null array of OSSL_PARAM values, terminated
-with a value containing a null B<key> member.
-B<print_values> is a control parameter, indicating that key values should be
+I<p> must be a non-null array of OSSL_PARAM values, terminated
```
Comment (t8m) [inline doc/man3/OSSL_PARAM_print_to_bio.pod:20]: Sorry for this nit, but this is still missing some markup `B<OSSL_PARAM>` and `B<BIO>`

File: doc/man3/OSSL_PARAM_print_to_bio.pod [hunk 41]
```diff
@@ -0,0 +1,42 @@
...
+I<print_values> is a control parameter, indicating that key values should be
+printed, in addition to key names.
+
+=head1 RETURN VALUES
+
+OSSL_PARAM_print_to_bio() returns 1 on success, and 0 on failure 
+
+=head1 HISTORY
+
+OSSL_PARAM_print_to_bio was added in OpenSSL 3.5
```
Comment (t8m) [inline doc/man3/OSSL_PARAM_print_to_bio.pod:31]: `OSSL_PARAM_print_to_bio()`

**Review by t8m — CHANGES_REQUESTED** [review]
Final two nits

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by nhorman** [comment]
merged to master, thank you for the reviews
